File: python/benchmark.py
#!/usr/bin/env python3
"""Benchmark experimental do Dijkstra (Python) — 30 rodadas por cenário."""

from __future__ import annotations

import argparse
import csv
import statistics
import time
from pathlib import Path

from dijkstra import dijkstra
from graphs import build_average_case, build_best_case, build_worst_case, edge_count

ROUNDS = 30
SIZES = {
    "pequeno": 300,
    "medio": 1500,
    "grande": 4000,
}

BUILDERS = {
    "melhor": build_best_case,
    "medio": build_average_case,
    "pior": build_worst_case,
}


def run_single(graph, source: int = 0) -> float:
    start = time.perf_counter()
    dijkstra(graph, source)
    return time.perf_counter() - start


def benchmark_case(case: str, size_label: str, n: int, rounds: int) -> dict:
    builder = BUILDERS[case]
    times = []
    edges = 0
    for r in range(rounds):
        graph = builder(n, seed=42 + r)
        if r == 0:
            edges = edge_count(graph)
        times.append(run_single(graph))

    return {
        "linguagem": "python",
        "caso": case,
        "tamanho": size_label,
        "n": n,
        "arestas": edges,
        "rodadas": rounds,
        "media_s": statistics.mean(times),
        "desvio_s": statistics.stdev(times) if len(times) > 1 else 0.0,
        "min_s": min(times),
        "max_s": max(times),
    }


def main() -> None:
    parser = argparse.ArgumentParser(description="Benchmark Dijkstra (Python)")
    parser.add_argument(
        "--output",
        type=Path,
        default=Path(__file__).resolve().parent.parent / "data" / "resultados_python.csv",
    )
    parser.add_argument("--rounds", type=int, default=ROUNDS)
    parser.add_argument("--quick", action="store_true", help="3 rodadas e tamanhos reduzidos")
    args = parser.parse_args()

    rounds = 3 if args.quick else args.rounds
    sizes = {"pequeno": 100, "medio": 400, "grande": 800} if args.quick else SIZES

    args.output.parent.mkdir(parents=True, exist_ok=True)
    rows = []
    for case in ("melhor", "medio", "pior"):
        for label, n in sizes.items():
            print(f"[Python] {case} / {label} (n={n}) — {rounds} rodadas...")
            rows.append(benchmark_case(case, label, n, rounds))

    fieldnames = list(rows[0].keys())
    with args.output.open("w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(rows)

    print(f"Resultados salvos em {args.output}")


if __name__ == "__main__":
    main()

File: javascript/benchmark.js
#!/usr/bin/env node
/** Benchmark experimental do Dijkstra (JavaScript) — 30 rodadas por cenário. */

import { performance } from "node:perf_hooks";
import { writeFileSync, mkdirSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { dijkstra } from "./dijkstra.js";
import {
  buildBestCase,
  buildAverageCase,
  buildWorstCase,
  edgeCount,
} from "./graphs.js";

const ROUNDS = 30;
const SIZES = {
  pequeno: 300,
  medio: 1500,
  grande: 4000,
};

const BUILDERS = {
  melhor: buildBestCase,
  medio: buildAverageCase,
  pior: buildWorstCase,
};

function mean(values) {
  return values.reduce((a, b) => a + b, 0) / values.length;
}

function stdev(values) {
  if (values.length < 2) return 0;
  const m = mean(values);
  const variance = values.reduce((s, x) => s + (x - m) ** 2, 0) / (values.length - 1);
  return Math.sqrt(variance);
}

function runSingle(graph, source = 0) {
  const start = performance.now();
  dijkstra(graph, source);
  return (performance.now() - start) / 1000;
}

function benchmarkCase(caseName, sizeLabel, n, rounds) {
  const builder = BUILDERS[caseName];
  const times = [];
  let edges = 0;

  for (let r = 0; r < rounds; r++) {
    const graph = builder(n, 42 + r);
    if (r === 0) edges = edgeCount(graph);
    times.push(runSingle(graph));
  }

  return {
    linguagem: "javascript",
    caso: caseName,
    tamanho: sizeLabel,
    n,
    arestas: edges,
    rodadas: rounds,
    media_s: mean(times),
    desvio_s: stdev(times),
    min_s: Math.min(...times),
    max_s: Math.max(...times),
  };
}

function toCsv(rows) {
  const headers = Object.keys(rows[0]);
  const lines = [headers.join(",")];
  for (const row of rows) {
    lines.push(headers.map((h) => row[h]).join(","));
  }
  return lines.join("\n");
}

function parseArgs() {
  const args = process.argv.slice(2);
  const opts = {
    output: join(dirname(fileURLToPath(import.meta.url)), "..", "data", "resultados_javascript.csv"),
    rounds: ROUNDS,
    quick: false,
  };
  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--output" && args[i + 1]) opts.output = args[++i];
    else if (args[i] === "--rounds" && args[i + 1]) opts.rounds = Number(args[++i]);
    else if (args[i] === "--quick") opts.quick = true;
  }
  return opts;
}

function main() {
  const opts = parseArgs();
  const rounds = opts.quick ? 3 : opts.rounds;
  const sizes = opts.quick
    ? { pequeno: 100, medio: 400, grande: 800 }
    : SIZES;

  const rows = [];
  for (const caseName of ["melhor", "medio", "pior"]) {
    for (const [label, n] of Object.entries(sizes)) {
      console.log(`[JavaScript] ${caseName} / ${label} (n=${n}) — ${rounds} rodadas...`);
      rows.push(benchmarkCase(caseName, label, n, rounds));
    }
  }

  mkdirSync(dirname(opts.output), { recursive: true });
  writeFileSync(opts.output, toCsv(rows), "utf-8");
  console.log(`Resultados salvos em ${opts.output}`);
}

main();
